---
title: "BINF 6210 Assignment 5"
author: "Amanda Meuser"
subtitle: "https://github.com/amanda-meuser/Binf6210Assign5"
date: "December 17th, 2021"
bibliography: library.bib
output: 
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: true
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, tidy.opts=list(width.cutoff=65), tidy = TRUE, error = TRUE)

```

\clearpage

# Introduction

Freshwater minnows are common globally, but with increasing anthropogenic disturbance world wide, species biodiversity is declining. Unique traits that these minnows possess may also be at risk of being lost to extinction. I will be examining the correlation between extinction status of the family *Cyprinidae* and their phylogenetic relatedness, while potentially expanding this to trait data, too. 

Do certain environments require specific conservation efforts, if there is one group of closely related species that are endangered which live in the same environment.



# Description of Data Sets

Data was downloaded from BOLD and IUCN Red List on Nov. 29th, 2021. From the IUCN Red List, data was retrieved from going to https://www.iucnredlist.org/search and searching the taxanomic family *Cyprinidae*. The `assessments.csv` file was used and renamed on my local computer for organization. From the BOLD database, data was retrieved by going to https://www.boldsystems.org/index.php/Public_SearchTerms and similarily searching *Cyprinidae*. The Combined TSV file was downloaded, which saved as a text file. 


# Code Section 1 â€“ Data Acquisition, Exploration, Filtering, and Quality Control

I will be using the `tidyverse` package [@Wickham2019], which includes `dplyr` for data wrangling and `ggplot2` for data visualization.
The package `phylosignal`, will be used to calculate the Lambda parameter of the phylosignal between cyprinid fish species and their IUCN Red List status [@Keck2016].

```{r, results='hide', message=FALSE}

# Package loading 
library(tools)
library(tidyverse)  # for manipulation of data.frames 
library(data.table) # for manipulation of data.tables
library(msa)        # for sequence alignment
library(Biostrings) # for creating DNAStringSet
library(seqinr)     # for computing distance matrix
library(ape)        # for creating basic dendrogram
library(wesanderson)# for plot colour palettes
#devtools::install_git("https://git.rud.is/hrbrmstr/waffle.git")
library(waffle)     # for waffle plot
library(magrittr)   # waffle dependency
library(phytools)   # for IUCN dendrogram
library(phylosignal)# for calculating phylosignal
library(phylobase)  # for creating phylo4d object
```



## Acquisition and Exploration

```{r, results='hide', message=FALSE, warning=FALSE}
# Importing CSV file from the IUCN Red List
dfFishStatus <- read_csv(file = "cyprinid_RedList_assessments.csv")

# Importing the text file from BOLD with sequence data
dfFishSeq <- read_delim("cyprinid_bold_data.txt", delim = "\t")

```
```{r, eval=FALSE}
# Comparing dimensions of the files
head(dfFishStatus)
summary(dfFishStatus)
```
```{r}
dim(dfFishStatus)
```
```{r}
dim(dfFishSeq)
```
```{r, eval=FALSE}
head(dfFishSeq)
summary(dfFishSeq)
summary(nchar(dfFishSeq$nucleotides))
```

There are almost 20 times as many observations with sequence data from BOLD than there are with conservation status from IUCN Red List. I will first do some individual quality control on the files, before merging them with a join function and removing some of the extra observations. 


## Filtering and Quality Control

### BOLD Data Set

First, I will filter the BOLD dataset, containing the sequence data, using the `data.table` package. After each filtering step, fewer observations remain. 

```{r}
dtFishSeq <- as.data.table(dfFishSeq)

# Remove observations with NA instead of a sequence
dtFishSeq <- dtFishSeq[complete.cases(dtFishSeq[, nucleotides])]

# Filter out sequences that aren't COI-5P
setkey(dtFishSeq, markercode)
dtFishSeq <- dtFishSeq["COI-5P"]

# Remove all sequences with dashes or Xs, with Ns at the beginning or end of the sequence, and Ns at a frequency greater than 2% in the sequence
dtFishSeq <- dtFishSeq[!nucleotides %like% "-+"][!nucleotides %like% "X+"][!nucleotides %like% "^N+"][!nucleotides %like% "N+$"]
dtFishSeq <- dtFishSeq[, gapN := str_count(nucleotides, c("[N]"))][, percentage_gapN := gapN/nchar(nucleotides)][!percentage_gapN > 0.02]

# Check that all of this worked
summDash <- dtFishSeq[, str_count(nucleotides, c("-"))] 
summary(summDash)
summX <- dtFishSeq[, str_count(nucleotides, c("X"))]
summary(summX)

(summary(dtFishSeq$percentage_gapN)*100) # format as a percent rather than a decimal
dtFishSeq <- dtFishSeq[, c("gapN", "percentage_gapN") := NULL]
rm(summDash, summX)

# Filter out sequences longer than 700bp or shorter than 400bp
dtFishSeq <- dtFishSeq[nchar(gsub("-", "", nucleotides)) %between% c(400, 700)]
summary(nchar(dtFishSeq$nucleotides))

# Check remaining species
dtFishSeq[, .N, by = species_name][order(-N)]

# Remove entries without species name data
dtFishSeq <- dtFishSeq[complete.cases(dtFishSeq[, species_name])]
dim(dtFishSeq)

```

I've noticed that some of the species names -- formatted *genus species* -- are actually just the genus name followed by 'sp.' or 'cf.' or 'aff.' and then a string of letters or numbers instead of a species name. As well, I noted a couple of hybrid crosses with X in between two species names, so I will be removing these, too. 

```{r, results='hide'} 
dtFishSeq <- dtFishSeq[!species_name %like% "sp."][!species_name %like% "cf."][!species_name %like% "aff."][!species_name %like% " X "]

summary(dtFishSeq)
unique(dtFishSeq$species_name)
```

Finally, I would like to remove empty columns from this data set, to declutter it and make it easier to work with, especially after joining with the IUCN Red List data.

```{r}
# Find all columns that only contain NA
nullNames <- dtFishSeq[, names(which(sapply(.SD, function(x) all(is.na(x)))))] 

# Remove these columns from the data table and create a duplicate of the data table
dtFishSeq2 <- dtFishSeq[, (nullNames) := NULL]
dim(dtFishSeq2)
rm(nullNames)
```

I need to reduce this data set so that I'm left with only one sequence per species, to match the IUCN data set, which has one Red List classification per species. 

```{r, results='hide'}  
unique(dtFishSeq2$nucleotides)
```

Using the `unique()` function above, I can see that there are 9262 unique sequences. I will first filter by this, to cut down on the size of the data set. 

```{r, results='hide'} 
distinctFishSeq <- dtFishSeq %>% 
  distinct(nucleotides, .keep_all = TRUE)

summary(distinctFishSeq)
unique(dtFishSeq2$nucleotides)
```

I now 9264 unique sequences. This is still more than 1 sequence per species, however, so I will remove/collapse the additional sequences for each species so that I only have 1 sequence per species. Using the `distinct()` function will simply keep the first entry for each unique entry in the species_name column. While this may not capture all sequence variation, it is an unbiased way to choose one individual to phylogenetically represent the species. 

```{r, results='hide'}
distinctFishSeq2 <- distinctFishSeq %>% 
  distinct(species_name, .keep_all = TRUE) 

unique(distinctFishSeq2$nucleotides)
dim(distinctFishSeq2)
```

I now have the same number of sequences as observations, so I know that I have one sequence per species. 

### IUCN Data Set

Now, I will filter the IUCN Red List data set. There's much less to filter for this dataset, as there is only one entry per species. 

```{r, results='hide'}
dtFishStatus <- as.data.table(dfFishStatus)

# Total species
dtFishStatus[, .N]

# Check number of each conservation status
dtFishStatus[, .N, by = redlistCategory][order(-N)]

# Remove Data Deficient enteries 
dtFishStatus <- dtFishStatus[!redlistCategory == "Data Deficient"]
dtFishStatus[, .N, by = redlistCategory][order(-N)] #check that it worked

# Move mislabeled observation of Red List Status into proper category
dtFishStatus$redlistCategory[dtFishStatus$redlistCategory == "Lower Risk/near threatened"] <- "Near Threatened"
dtFishStatus[, .N, by = redlistCategory][order(-N)] #check that it worked

# Subset this data set to keep only relevant columns containing species name and conservation status.
dtFishStatus2 <- dtFishStatus[, c("scientificName", "redlistCategory"), with = F]

# Assign a numerical rank to IUCN Red List category, for later plotting
dtFishStatus2$redlistRank <- as.numeric(factor(dtFishStatus2$redlistCategory, levels = c("Least Concern", "Near Threatened", "Vulnerable", "Endangered", "Critically Endangered", "Extinct in the Wild", "Extinct"))) 

```

### Combined Data Set

```{r, results='hide'}
# Merge data sets with data.table
dtFishJoin <- merge(distinctFishSeq2, dtFishStatus2, by.x = "species_name", by.y = "scientificName", all.x = FALSE, all.y = FALSE)

class(dtFishJoin)
dim(dtFishJoin)
head(dtFishJoin)
names(dtFishJoin)
```

There are 965 species in my combined data set, however, as we were advised to have about 50-100 species for this project, I will constrict my data set based on country, to keep only species found in North America. Additionally, for whatever reason, 2 species have very short, negative branch lengths later on when I create my simple dendrogram, so I've been advised by Jacqueline to just remove these two.

```{r, results='hide'}
# Look at the number of species per country
(country <- table(dtFishJoin$country))

# Filter to keep only species from Canada, United States, or Mexico, and remove 2 observations
dtFishJoinNA <- dtFishJoin[country %like% "Canada|United States|Mexico"][!species_name %like% "Lythrurus fumeus|Pteronotropis hypselopterus"]

dim(dtFishJoinNA)
```

192 observations is a far easier amount of data to work with than the original 965. I will keep these samples to retain as much variability in the data as possible, and only down sample further if needed for visualizations. I have shown, in Fig. 1, that I primarily have species of "Lease Concern" from the IUCN Red List, but that there are a small amount of species that are threatened or endangered. 

Next, I will align my sequences using `msa()`. I chose to use the default ClustalW algorithm over the well-known MUSCLE algorithm, as ClustalW is comparable in alignment score to MUSCLE and while MUSCLE is faster with large quantities of sequences or with long sequences, my data set consists of only 100 short sequences [@Edgar2004]. Thus, the extra computational power is not needed and ClustalW will perform adequately for this assignment. I chose Neighbour-Joining (NJ) method for clustering, as the secondary option UPGMA has been known to crash the R session on Windows, and NJ has been successfully used for fish species in the literature [@Zou2020]. I used the default parameters for all other arguments. 

```{r}
# Create a subset of the data frame (just identifier and sequence)
FishSeqsNA <- dtFishJoinNA[, c("species_name", "nucleotides")]

# Convert data table/frame to a DNA string set
FishSeqsNA <- DNAStringSet(FishSeqsNA$nucleotides)
class(FishSeqsNA)

# Add names to DNAStringSet
names(FishSeqsNA) <- paste(dtFishJoinNA$species_name)
metadata(FishSeqsNA)

# Sequence alignment
AlignedFishSeqsNA <- msaClustalW(FishSeqsNA, cluster = "nj")
AlignedFishSeqsNA

# msaPrettyPrint(AlignedFishSeqs, file=tmpFile, output="pdf", y=c(164, 213),
#                subset=c(1:6), showNames="left", showLogo="top", 
#                logoColor="rasmol", showConsensus = "none", showLegend=FALSE,
#                askForOverwrite=FALSE)

```

I have visualized the sequence alignment as a simple dendrogram, in Fig. 2. There appear to be no outlier sequences, so I will continue with my main analysis. 


# Main Software Tools Description

I chose to use the package `phylosignal` as my main software tool [@Keck2016]. I chose this package for a couple of reasons. Aside from meeting my main criterion of being able to calculate the Lambda parameter for a trait, I primarily chose `phylosignal` because it contained many functions for visualizing the output from the phylosignal calculation, done with the `phyloSignal()` function. Particularly, I utilized the `phyloCorrelogram()` function to create a correlogram which displays the correlation between phylogenetic distance and IUCN Red List status. Additionally, the `phylosignal` package uses phylogenetic trees of the object type "phylo", which is a common type that makes it compatible with other phylogenetic R packages, such as `ape`, `phytools`,



# Code Section 2 - Main Analysis

## Calculation of Phylosignal

```{r}
# Converting the alignment to a seqinr object
ConvertAlignNA <- msaConvert(AlignedFishSeqsNA, type="seqinr::alignment")

# Using a function from the package seqinr to create a distance matrix of the aligned sequences
DistMatrixNA <- dist.alignment(ConvertAlignNA, "identity")

# Using a function from the package ape to create a Neighbour-Joining Tree
FishTreeNA <- nj(DistMatrixNA)

# Ensure that it's in "phylo" format
class(FishTreeNA)

# Isolate just the ranks, indicating Red List Category
ranksNA <- dtFishJoinNA$redlistRank

# Assign species names to their rank
ranksNA <- structure(ranksNA, names = (dtFishJoinNA$species_name))
ranksNA <- data.frame(ranksNA)

# Create phylo4d object
phyloSigFish <- phylo4d(FishTreeNA, ranksNA)

# Calculate Lambda parameter for phylosignal
signal <- phyloSignal(p4d = phyloSigFish, method = "Lambda")

(signal)

```


# Results and Discussion





# Figures

## Fig 1: Barplot of species, per IUCN status, per country

```{r, fig.show='hold', fig.cap = "Barplot of number of species per ICUN Red List classification. 1 box = ~2 species, n = 192."}

# # Define colour palette 
 plotCol <- wes_palette("Zissou1", 5, type = "continuous")

FishWaffleNA <- dtFishJoinNA %>%
  dplyr::count(redlistRank)

# Make waffle plot
ggplot(FishWaffleNA, aes(fill = redlistRank, values = n)) +
  geom_waffle(n_rows = 10, size = 0.33, colour = "white", flip = T, make_proportional = T) +
  scale_fill_manual(name = "IUCN Conservation \nStatus", values = plotCol, labels = c("Least Concern", "Near Threatened", "Vulnerable", "Endangered", "Critically Endangered")) +
  coord_equal() +
  theme_void() 

```


## Fig 2: Dendeogram 


```{r, fig.show='hold', fig.cap = "Phylogenetic tree of North American cyprinid species. Tip labels have been removed for clarity. n = 194"}

# Create dendrogram, using same colour palette as waffle plot
plot(FishTreeNA, edge.color = plotCol[1], tip.color = "white", cex = 0.05)

```


## Fig 3: Dendrogram coloured by ICUN status

```{r}
# Move only the Least Concern IUCN status into a separate data frame
onlyLC <- dtFishJoinNA[redlistCategory %like% "Least Concern"]

# Move all other categories into another data frame
others <- dtFishJoinNA[!redlistCategory %like% "Least Concern"]

# Sample down only the Lease Concern so that there's 30 
set.seed(20)
onlyLC <- sample_n(onlyLC, 30, replace = F)

# Combine the two data frames
dtFish60 <- rbind(others, onlyLC)

dim(dtFish60)
table(dtFish60$redlistCategory)
class(dtFish60)
```

There are now only 60 observations in this data frame/table, so it will be easier to make into a dendrogram. First, I need to align only these 60 sequences, and ensure that they are an object of class "phylo", so that they can be plotted. 

```{r}
# Create a subset of the data frame (just identifier and sequence)
FishSeqs60 <- dtFish60[, c("species_name", "nucleotides")]

# Convert data table/frame to a DNA string set
FishSeqs60 <- DNAStringSet(FishSeqs60$nucleotides)
class(FishSeqs60)

# Add names to DNAStringSet
names(FishSeqs60) <- paste(dtFish60$species_name)
metadata(FishSeqs60)

# Sequence alignment
AlignedFishSeqs60 <- msaClustalW(FishSeqs60, cluster = "nj")

# Converting the alignment to a seqinr object
ConvertAlign60 <- msaConvert(AlignedFishSeqs60, type="seqinr::alignment")

# Using seqinr to create a distance matrix from the aligned sequences
DistMatrix60 <- dist.alignment(ConvertAlign60, "identity")

# Using ape to create a Neighbour-Joining Tree
FishTree60 <- nj(DistMatrix60)
class(FishTree60)

```

```{r, fig.show='hold', fig.cap = "Phylogenetic tree of North American cyprinid species and their status on the IUCN Red List. Cool to warm colouring indicates increasing risk of species loss; 1 = Least Concern status and 5 = Critically Endangered status. Note: number of species of Least Concern status have been down sampled from 164 to 30, for readability of the plot. n = 60."}
# Isolate just the ranks, indicating Red List Category
ranks60 <- dtFish60$redlistRank

# Assign species names to their rank
ranks60 <- structure(ranks60, names = (dtFish60$species_name))

# Create the dendrogram
tree <- contMap(FishTree60, ranks60, plot=FALSE)
tree <- setMap(tree, invert = TRUE)

# Plot the dendrogram
plot(tree, fsize = c(0.5,1), outline = FALSE, lwd = c(3,7), leg.txt = "Red List Category")

```


## Fig 4: Correlation of IUCN status and phylogeny

```{r, fig.show='hold', fig.cap = "Correlation between phylogenetic distance and IUCN Red List status. Red on x-axis indicates statistically significant correlation."}
# Create phylocorrelogram object
phyloSigFish.crlg <- phyloCorrelogram(phyloSigFish)

# Create plot
plot(phyloSigFish.crlg, mar=c(10,3,2,2))
```


# Acknowledgements


# Bibliography

## Websites


Martin, A., & akrun. (2020, November 2). What is the most efficient way to remove empty columns in a   datatable in R. Stack Overflow. Retrieved December 10, 2021, from                              https://stackoverflow.com/questions/64759969/what-is-the-most-efficient-way-to-remove-empty-columns-in-a-datatable-in-r 

Ram, K. (2015, January 22). Karthik/Wesanderson: A Wes Anderson color palette for R. GitHub. Retrieved December 16, 2021, from https://github.com/karthik/wesanderson 

Rudis, B. (2015, March 18). HRBRMSTR/Waffle: Make Waffle (square pie) charts in R. GitHub. Retrieved  December 12, 2021, from https://github.com/hrbrmstr/waffle 

Schork, J. (2021, June 25). Select subset of data table columns in R (example): Keep variables.       Statistics Globe. Retrieved December 12, 2021, from                                            https://statisticsglobe.com/select-subset-of-data-table-columns-in-r 

VanHespen, R. (2021, January 14). Writing your thesis with R Markdown (2) â€“ Text, citations and        equations. Rosanna's Research. Retrieved December 16, 2021, from                                        https://rosannavanhespen.nl/rmarkdown/writing-your-thesis-with-r-markdown-2-text-citations-and-equations/ 


## 9.2 Journal Articles



